### PR TITLE
fix(layout): give the footnote band its own column layout and separator placement

### LINKED ISSUE #3960: Footnote band in a Hebrew multi-column section: w15:footnoteColumns ignored, separator pinned to the left
## What happened?

A Hebrew (RTL) Word document with three sections — single-column, two-column `continuous`, single-column `continuous` — where all three `sectPr` declare `<w15:footnoteColumns w:val="1"/>`.

Word lays the footnotes out as **one strip across the whole content area**, with the short `w:separator` rule at that strip's **right** edge. SuperDoc lays them out **one body column wide**, with the rule at that column's **left** edge.

These are two independent defects that happen to surface together.

### 1. `w15:footnoteColumns` is never read, so the band reuses the body's `w:cols`

Geometry read out of the DOM in Chrome against the packaged dist (`data-layout-story` identifies the note band), for A4 with 1800tw side margins and a 708tw gutter:

| Element | x | width |
|---|---|---|
| body paragraph, single-column section | 516 | 554 |
| body paragraph, column 0 | 516 | 253 |
| body paragraph, column 1 | 816 | 253 |
| note band (`data-layout-story="footnote:*"`) | 516 | 253 |
| **what Word draws** | **516** | **554** |

The numbers follow from the XML exactly: content area `11906 − 1800 − 1800 = 8306tw = 554px` at 96dpi; one body column `(8306 − 708) / 2 = 3799tw = 253px`.

A second symptom falls out of the same reuse: the one note measurement width is the **minimum body column width across the whole document**, so a single two-column section in the middle narrows the notes of every single-column section as well.

### 2. The separator rule is pinned to the note column's left edge

`<w:separator/>` is not engine decoration — it is a **run**, inside its own paragraph in `footnotes.xml` (`w:footnote w:type="separator"`, id `-1`). Its horizontal placement is therefore that paragraph's resolved inline direction, `w:jc` and `w:ind`.

In this document that paragraph declares no direction of its own:

```xml
<w:footnote w:type="separator" w:id="-1">
  <w:p><w:pPr><w:spacing w:after="0" w:line="240" w:lineRule="auto"/></w:pPr>
    <w:r><w:separator/></w:r></w:p>
</w:footnote>
```

and inherits from `w:docDefaults/w:pPrDefault`, which carries `<w:bidi/>` and `<w:jc w:val="both"/>`. Word draws the rule on the right.

Worth stating explicitly, because it is the tempting wrong fix: the **section's** `w:bidi` is not the signal here. The axes are independent (§17.6.1 for the section, §17.3.1.6 for the paragraph), and they do come apart — a Hebrew section whose separator paragraph resolves LTR gets a rule on the **left** in Word, and an LTR section whose separator paragraph resolves RTL gets one on the right. Keying the placement on the section direction renders both of those backwards.

## Steps to reproduce

No attachment needed; the relevant markup is small enough to build a synthetic document from.

1. `word/document.xml` — three sections, the middle one two-column, `w:bidi` throughout, all three declaring a one-column note band:

```xml
<w:sectPr>
  <w:footnotePr><w:numFmt w:val="hebrew1"/></w:footnotePr>
  <w:type w:val="continuous"/>
  <w:pgSz w:w="11906" w:h="16838"/>
  <w:pgMar w:top="1440" w:right="1800" w:bottom="1440" w:left="1800"
           w:header="708" w:footer="708" w:gutter="0"/>
  <w:cols w:num="2" w:space="708"/>
  <w:bidi/>
  <w15:footnoteColumns w:val="1"/>
</w:sectPr>
```

2. `word/styles.xml` — an RTL paragraph default, which is what a Hebrew document normally carries:

```xml
<w:docDefaults><w:pPrDefault><w:pPr><w:bidi/><w:jc w:val="both"/></w:pPr></w:pPrDefault></w:docDefaults>
```

3. `word/footnotes.xml` — the default separator paragraph as Word writes it (shown above), plus one or more footnotes anchored in **both** body columns of the middle section.
4. Open in SuperDoc, open the same file in Microsoft Word, and compare the note area on the page holding the two-column section.

Expected (Word): one note strip spanning the content area, its rule at the right edge.
Actual: a strip one body column wide, its rule at that column's left edge.

## SuperDoc version

`superdoc@2.11.1-next.1` / `@superdoc/docx-engine@0.10.1-next.1`

## Browser

Chrome

## Additional context

- Display only, not data loss: an open→export round trip returns `w15:footnoteColumns` unchanged in all three `sectPr`.
- The element is a Word 2012 extension (`http://schemas.microsoft.com/office/word/2012/wordml`) and appears in `mc:Ignorable`, so skipping it is well-formed — just not faithful.
- Fixing this needs both halves of the stack: the layout side (the band's own column layout, the measurement width, the reference→stack mapping, the reserve, and the separator's x) and the importer side (reading `w15:footnoteColumns` off `sectPr`, and resolving the separator paragraph's `w:pPr` through its style chain). A PR for the layout half follows this issue.
- Happy to split this into two issues if you would rather triage the band geometry and the separator placement separately.

### PR DESCRIPTION
**Pull request by Nathaniel-260**
Fixes #3960

> [!IMPORTANT]
> **Depends on #3953 and cannot merge before it.** `main` has no column direction at all today;
> #3953 is what gives `ColumnLayout` its `direction` and mirrors the strip in RTL, and the footnote
> band inherits that axis. Merging this first would not compile against `main`.
>
> **Two commits are mine** — `fix(layout): give the footnote band its own column layout and
> separator placement` and `docs(layout): state the footnote band's separator invariants
> accurately`. Everything below them in the diff belongs to #3953 and disappears once it merges.
> `CONTRIBUTING.md` asks for `main`, so the base is `main` and the stack shows up in the diff.

A Hebrew (RTL) document with a two-column `continuous` section and `<w15:footnoteColumns w:val="1"/>` on every `sectPr` printed its footnotes in a strip one body column wide, with the separator rule on the wrong side of it. Word prints one strip across the content area, rule at the right.

Geometry read out of the DOM in Chrome against the packaged dist, before the change (`data-layout-story` identifies the band):

| Element | x | width |
|---|---|---|
| body paragraph, single-column section | 516 | 554 |
| body paragraph, column 0 | 516 | 253 |
| body paragraph, column 1 | 816 | 253 |
| note band | 516 | 253 |

A4 with 1800tw side margins and a 708tw gutter: content area `8306tw = 554px` at 96dpi, one body column `(8306 - 708) / 2 = 3799tw = 253px`. After the change the band is emitted at the content-area origin across its full width (`x = marginLeft`, `width = 553.73px`), with both columns' notes in one stack in note order. That is asserted directly on the layout in `footnoteBandColumns.test.ts` rather than eyeballed in a browser.

One visible consequence worth naming: the `w:separator` rule is `0.5x` the note column's width, which is pre-existing, so a full-width band gets a proportionally longer rule -- 277px rather than 126px here. That follows from the band being wider and is not something this PR changes; if Word's rule is a fixed length instead, that is a separate fix to the factor.

Two independent defects. They are in one commit because the shared files carry hunks from both, and a hand-split would have produced a half-broken intermediate; the sections below and the commit message separate them, and I am happy to split the PR itself in two if you would rather review them apart.

### 1. The band reused the body's `w:cols`

`w15:footnoteColumns` was never read anywhere in the tree, and every note-plane stage took its geometry from the body. `resolveFootnoteBandColumns` is now the single place that decision is made, and four stages moved onto the band's plane together, because a disagreement between any two of them is visible:

- **Measurement width** — notes were measured at a body column's width, and specifically at the *minimum* body column width anywhere in the document, so one two-column section in the middle narrowed the notes of every single-column section too. It is now the narrowest **band** in the document.
- **Reference → stack assignment** — keyed by the body column the anchor sits in; now mapped through the band's count, so several body columns can feed one stack, in note order.
- **The planner and its reserve** — a merged band is one stack whose height is the *sum* of the per-column stacks, not their max, so the old reserve let the band run past the bottom margin.
- **The painted x/width**, plus the retained-band and coupled-pagination paths and the warm seed.

Absent the element, every path is byte-identical to before.

### 2. The separator rule was pinned to the column's left edge

`<w:separator/>` is a run inside its own paragraph in `footnotes.xml`, so its placement is that paragraph's resolved inline direction, `w:jc` and `w:ind`.

It is deliberately **not** keyed on the section's `w:bidi`, which is the tempting wrong fix. The axes are independent (§17.6.1 vs §17.3.1.6, as `direction-context.ts` already states) and they come apart: a Hebrew section whose separator paragraph resolves LTR gets a rule on the **left** in Word, and the reverse for an LTR section with an RTL separator paragraph. Nothing about the section reaches `resolveFootnoteSeparatorX` — its inputs are the note column's geometry and that one paragraph — and both divergent cases have tests. `w:jc` resolves exactly the way `resolveTextAlign` already resolves it for text.

## What this PR does not do

The importer half, which is where the two new inputs get populated:

| Input | Source |
|---|---|
| `SectionBreakBlock.footnoteColumns`, `LayoutOptions.footnoteColumns` | `w:sectPr/w15:footnoteColumns/@w:val` (`0`/absent = match the body) |
| `footnotes.separatorParagraph`, `footnotes.continuationSeparatorParagraph` | the resolved `w:pPr` of `w:footnote w:type="separator"` / `"continuationSeparator"`, through the style chain — a separator paragraph usually declares nothing and inherits from `w:docDefaults/w:pPrDefault` |

Until they are populated, documents render exactly as they do today: the band resolves to the body layout it already used, and the separator keeps its historical left edge. The `w15:footnoteColumns` reader added to `to-flow-blocks.ts` is the reference implementation, prefix and `0` default included.

## Design decisions worth a look

- **A band with more columns than the body clamps back to the body.** The note planner carries a column's overflow to the same column on the next page, never sideways, so honouring a wider count would paint a half-width strip beside an empty one and push notes onto later pages — strictly worse than the body-matching band the default produces. Rationale is in the helper's doc comment.
- **Note measurement stays one global pass.** `footnoteConstraints` is a single object and a note's section is only known after the body is paginated, so the width is the narrowest band in the document. Painting clamps to `min(bandColumnWidth, measurementWidth)`, so measurement and paint can never disagree; a document mixing band widths under-fills its wider bands rather than spilling out of its narrower ones. Per-section note measurement is the follow-up that removes the compromise.
- **`w:ind` narrows the separator's extent; `firstLine`/`hanging` are not read.** They move the first line of a text paragraph and I have no evidence Word applies them to a lone separator run — better handled once the behaviour is known than guessed at.

## Testing

44 new tests across 4 files, including the exact `sectPr` set of the document that surfaced this, rebuilt synthetically (no `.docx` fixture added).

- `npx vp test run --root packages/layout-engine/contracts` — 32 files, 538 pass
- `npx vp test run --root packages/layout-engine/layout-bridge` — 96 files, 1895 pass
- `npx vp test run --root packages/layout-engine/tests` — 284 pass, plus one pre-existing failure noted below
- `npx tsc -b tsconfig.references.json` — clean
- `vp lint` on the touched trees — no errors; `vp fmt` applied
- `node scripts/check-public-boundary.mjs` — OK

Two things I could not run in my checkout, both unrelated to this change: `pnpm test` at the root (the root `prepare` script fails in my shell) and anything scoped to `packages/superdoc` (installed engine `0.10.0-next.12` vs the declared `0.10.1-next.1`). Also pre-existing and untouched here: `architecture-boundaries.test.ts` Guard E flags `getBoundingClientRect` in `painters/dom/src/ruler/ruler-renderer.ts`.

No public API surface was grown — the new helpers are exports of `@superdoc/contracts`, which is a private workspace package and is not re-exported from `superdoc`'s public index, so no `tests/consumer-typecheck/` fixture applies.

### Verified through the paint path, not just the layout

Because a layout assertion does not prove anything reaches the screen, I also ran the full `incrementalLayout` -> resolve -> `DomPainter` pipeline and read `style.left` / `style.width` back off the painted elements. Measured, in CSS px:

| Case | Measured | Expected |
|---|---|---|
| merged band, both notes | left `120`, width `553.73` | full content area |
| merged band, note order | note 1 top `1004.53` < note 2 top `1016.53` | note order preserved |
| merged band, separator | width `276.87`, left `396.87` | half the band, at its right edge |
| `footnoteColumns` omitted | two strips of width `253.27`, two separators | today's behaviour, unchanged |
| tall notes (2 x 200px, merged) | note tops `624.53` and `826.53`, gap `202` | the stack is the SUM, not the max |
| tall notes, band bottom | `1026.53` vs the page limit `1026.53` | inside the bottom margin, bottom-anchored |
| RTL section, **LTR** separator paragraph | left `120` | the LEFT edge |
| LTR section, **RTL** separator paragraph | left `396.87` | the RIGHT edge |

The last two are the pair that a section-direction fix would get backwards, so they are measured rather than argued.

The painter itself was checked for the same class of bug and is clean: `applyFragmentFrame` / `applyResolvedFragmentFrame` write `style.left` / `style.width` straight from the fragment's own `x` / `width` and never re-derive them from `page.columns`.

One limit worth naming: text fragment wrappers get no explicit `style.height`, so the "band bottom" figures above are a painted top read from the DOM plus the measured line height, not a second independent DOM measurement.

No screenshot: the change is geometry, and it is pinned numerically instead.

## Stacking

Repeated from the top because it decides merge order: this is built on #3953, which is what gives `ColumnLayout` its `direction` and mirrors the strip in RTL. `main` has no column direction at all, so the band has no axis to inherit and this does not stand alone -- #3953 has to land first. The two commits named at the top are mine; everything below them belongs to that PR and disappears from this diff once it merges.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/superdoc/docx-editor/pull/3961?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/layout-engine/contracts/src/footnote-band-columns.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * columns, which the page painter derives from `page.columns`; the band draws its own horizontal
+ * `w:separator` and no vertical rules of its own.
+ */
+export function resolveFootnoteBandColumns(
+  bodyColumns: ColumnLayout | undefined,
+  footnoteColumns: number | undefined,
+): ColumnLayout {
+  const body = cloneColumnLayout(bodyColumns);
+  const count = resolveFootnoteColumnCount(bodyColumns, footnoteColumns);
+  if (count >= resolveColumnCount(bodyColumns)) return body;
```
[the remark this answers is not in the document]
Comment (Nathaniel-260) [reply packages/layout-engine/contracts/src/footnote-band-columns.ts:62]: Checked, and the premise is right but the conclusion inverts the branch's purpose, so I fixed the comment instead of the code (fe8d7b6).

`withSeparator` really is preserved here — `cloneColumnLayout` copies it. But this branch's contract is "the body layout, verbatim", and that is precisely what makes every document *without* `w15:footnoteColumns` render byte-identically to before the change. Stripping a field to make the returned object differ from the body layout would trade that guarantee for nothing.

Nothing, because the field is inert on this plane:

- `@w:sep` drives the VERTICAL rules between body columns, and the painter derives those from `page.columns` / `page.columnRegions` (`renderer.ts:1800`), never from the band layout.
- No footnote consumer reads the band geometry's `separatorX`. The band calls only `getColumnX` and `getColumnWidth`; `grep -n 'separatorX\|getColumnSeparatorPositions' incrementalLayout.ts` is empty.

The merge branch does drop it, and a test asserts that. What was actually wrong was the doc comment claiming it "is never carried over" — a blanket statement true of only one of the two branches. It now states which branch drops it, which keeps it, and why keeping it is inert.

File: packages/layout-engine/layout-bridge/src/incrementalLayout.ts [hunk 2]
```diff
@@ -4076,7 +4188,18 @@ export async function incrementalLayout(
...
-                x: columnX,
+                // The mark is a run in its own paragraph in footnotes.xml, so it sits at that
+                // paragraph's start edge — the RIGHT of the note column when that paragraph resolves
+                // RTL. A full-width continuation mark fills the extent and lands on columnX either
+                // way; the short `w:separator` is the one this moves.
+                x: resolveFootnoteSeparatorX({
+                  columnX,
+                  columnWidth: contentWidth,
+                  separatorWidth,
+                  attrs: isContinuation
```
[the remark this answers is not in the document]
Comment (Nathaniel-260) [reply packages/layout-engine/layout-bridge/src/incrementalLayout.ts:4204]: Valid, and it was already the known gap — but the comment stated only half of it, so I completed it (fe8d7b6).

The reuse path is gated, though not by the plan-input comparisons: `retainedFootnotePlaneBaseAdoptable` requires `layoutReuse.dependencyProof.renderInputsUnchanged === true` before it adopts a previous page's band fragments. A resolved separator paragraph changing means footnotes.xml, a style, or `w:docDefaults` changed, which is a render-input change — so an honest host proof already forces recomputation, and no invalidation of my own would add anything on top of a proof that is telling the truth.

What was missing was saying so. The field now documents that reuse safety rests on that proof, that an adopted band keeps the rule's old x, and that a host which starts populating the field **must** include it in the proof.

I deliberately did not add a separate signature comparison for it. The field is unpopulated today — the importer half is not in this PR — so the path is unreachable, and speculative invalidation machinery for an input nobody feeds yet is exactly the kind of thing this repo's review rules ask contributors not to add. The requirement is instead recorded where the person wiring it up will read it. If you would rather have the belt-and-braces check in the gate, say so and I will add it with a test that drives it.

File: packages/layout-engine/contracts/src/column-layout.ts [hunk 3]
```diff
@@ -242,13 +309,53 @@ export function getColumnSeparatorPositions(geometry: ColumnGeometry[], originX
...
+ * caller keeps working unchanged.
+ */
 export function getColumnAtX(geometry: ColumnGeometry[], x: number, originX = 0): number {
   if (geometry.length === 0) return 0;
   const cx = x - originX;
+  const mirrored = geometry.length > 1 && geometry[1].x < geometry[0].x;
   let result = 0;
   for (const col of geometry) {
-    if (cx >= col.x) result = col.index;
+    if (mirrored ? cx <= col.x + col.width : cx >= col.x) result = col.index;
```
[the remark this answers is not in the document]
Comment (Nathaniel-260) [reply packages/layout-engine/contracts/src/column-layout.ts]: You are right and my first read of this was wrong. It is a real defect, not a boundary-convention preference.

I initially assessed it as a false positive on the grounds that `cx <= col.x + col.width` is the exact mirror of the LTR branch's `cx >= col.x`: both make a shared edge belong to the column whose span *starts* there, and for a **click** that is true.

What I missed is that no caller passes a click. Every one of the seven `determineColumn` call sites in `position-hit.ts` passes `fragment.x`, and `column-balancing.ts` passes `fragment.x` too — a fragment **origin**, which is the column's physical LEFT edge in both directions. With `w:space="0"` (ECMA-376 section 17.6.3) adjacent columns share an edge, so in an RTL section `col0.x == col1.x + col1.width` exactly, and the inclusive form let column 1 claim column 0's own origin. Two columns over 602px mirror to col 0 `[301,602)` and col 1 `[0,301)`: `findColumnContaining(301)` answered 0 and `getColumnAtX(301)` answered 1, so the two resolvers disagreed at the one point they can be made to disagree about — and a caret in the first column of a zero-gutter Hebrew section landed in the second.

With a gutter the same bound over-claimed a column's trailing edge, which is gutter and belongs to the column preceding it in fill order, contradicting this function's own documented rule.

`cx <` is correct on both counts and makes the two resolvers agree everywhere they can both answer. Confidence 9 was warranted.

**Fixed one PR up the stack, not in this one.** The change landed in `fix/column-gaps-and-balance-order` (#3962) as `160c3c2`, with a test that pins the zero-gutter boundary and the trailing-edge case in both resolvers. That means **#3953 and this PR will merge with the defect still present** and #3962 fixes it immediately after. I would rather say that plainly than split an already-verified commit to backdate it — but if you would prefer the fix in #3953 so nothing merges with a known wrong answer, say so and I will move it.

layout-bridge, which owns `determineColumn`, is 96 files / 1895 pass against the change; contracts is 32 files / 543 pass.

File: packages/layout-engine/layout-bridge/test/position-hit.test.ts [hunk 4]
```diff
@@ -111,6 +111,35 @@ describe('determineColumn (SD-2629: resolved per-column boundaries)', () => {
...
+    // the wrong column — the issue's "clicks will select the wrong column".
+    const columns = { count: 3, gap: 24, direction: 'rtl' as const };
+    const page = {
+      columns,
+      margins: { left: 96, right: 96 },
+      size: { w: 816, h: 1056 },
+    } as unknown as Page;
+    const layout = { pageSize: { w: 816, h: 1056 }, columns, pages: [page] } as unknown as Layout;
+
+    // Content width 624 -> 192px columns. Mirrored, column 0 spans 528..720, column 1 336..528,
```
[the remark this answers is not in the document]
Comment (Nathaniel-260) [reply packages/layout-engine/layout-bridge/test/position-hit.test.ts]: Correct, and fixed — the comment was a full gutter off.

Recomputed from the real geometry: content width 624 over 3 columns with a 24px gutter gives 192px columns, mirrored to absolute 528..720, **312..504**, 96..288, with the gutters at 504..528 and 288..312. Columns 0 and 2 in the old comment were right; column 1 read 336..528, which is `col1.x + gap` rather than `col1.x`.

The assertions were unaffected — the probe at 400 falls inside column 1 either way — so this only ever misled a reader, which is exactly why it was worth fixing rather than closing. The comment now derives each start from the previous one as `width + gap` and names both gutters, so the next person can check it rather than trust it.

Fixed in `160c3c2`, the same commit as the `getColumnAtX` boundary fix.

[the remark this answers is not in the document]
Comment (Nathaniel-260) [reply packages/layout-engine/layout-bridge/test/position-hit.test.ts]: Correction to my earlier reply, which said this was fixed in `160c3c2`. That is true, but `160c3c2` is on #3962 — **not on this branch**, and not on #3953 either.

Verified just now against the pushed tips:

| branch | `position-hit.test.ts:126` |
|---|---|
| #3953 `fix/rtl-column-order` | `column 1 336..528` (still wrong) |
| #3961 this PR | `column 1 336..528` (still wrong) |
| #3962 `fix/column-gaps-and-balance-order` | `column 1 312..504` (fixed) |

So the misleading comment you flagged is still present in the diff you are reading. The assertions are unaffected — the probe at 400 falls inside column 1 either way — but the comment is a full gutter off, exactly as you said.

The line originates in #3953, so the clean place to fix it is there, where it would then reach both children on rebase. Leaving this open rather than resolving it, because on this branch the defect is still live. Tell me if you would rather I put the one-line fix into #3953 and I will.

File: packages/layout-engine/contracts/src/column-layout.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nathaniel-260) [reply packages/layout-engine/contracts/src/column-layout.ts]: Moved the fix down into #3953, so this PR no longer carries the defect and this thread is a closed loop.

#3953 is where the mirrored branch and its inclusive bound are introduced, so that is where the defect enters the tree. Fixing it two PRs later meant #3953 and #3961 would both merge with a line already known to be wrong. It is `ad9bde6` there now, and this branch is rebased onto it.

Only the boundary fix moved. The commit on #3962 that used to carry it also contained an unrelated change to `columnRenderLayoutsEqual`; that stayed where it was and has been renamed to describe what it actually does, since it no longer touches this bound.

Verified the move changed nothing but placement: #3962's final tree is byte-identical to what it was before, `git diff` between the two tips is empty. Suites green at 2399 passed / 1 skipped across 125 files.

File: packages/layout-engine/layout-bridge/test/position-hit.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nathaniel-260) [reply packages/layout-engine/layout-bridge/test/position-hit.test.ts]: Moved the fix down into #3953, so this PR no longer carries the defect and this thread is a closed loop.

#3953 is where the mirrored branch and its inclusive bound are introduced, so that is where the defect enters the tree. Fixing it two PRs later meant #3953 and #3961 would both merge with a line already known to be wrong. It is `ad9bde6` there now, and this branch is rebased onto it.

Only the boundary fix moved. The commit on #3962 that used to carry it also contained an unrelated change to `columnRenderLayoutsEqual`; that stayed where it was and has been renamed to describe what it actually does, since it no longer touches this bound.

Verified the move changed nothing but placement: #3962's final tree is byte-identical to what it was before, `git diff` between the two tips is empty. Suites green at 2399 passed / 1 skipped across 125 files.
